### ISSUE TITLE
DNS lookup from aks pods takes more than 5 seconds

### ISSUE THREAD

**Issue by vishnujanardhanan** [Azure/AKS#1326]
**What happened**: DNS lookup from aks pods taken more than 5 seconds intermittently 

**What you expected to happen**: DNS lookup should be quick

**How to reproduce it (as minimally and precisely as possible)**: ran a curl https://www.google.com and can see the name lookup taking time in curl's output. Attaching the curl output file.

**Anything else we need to know?**:

**Environment**:aks cluster
- Kubernetes version (use `kubectl version`):1.13.12
- Size of cluster (how many worker nodes are in the cluster?): tested in 2 node and 3 node cluster.
- General description of workloads in the cluster (e.g. HTTP microservices, Java app, Ruby on Rails, machine learning, etc.): spring app, microservices. able to replicate the issue via curl
- Others:
[curl_google_from_pod.txt](https://github.com/Azure/AKS/files/3865246/curl_google_from_pod.txt)

<img width="1675" alt="Screenshot 2019-11-20 at 8 21 01 AM" src="https://user-images.githubusercontent.com/31517838/69205112-09170380-0b6f-11ea-8153-5bc72dfdce74.png">
<img width="1675" alt="Screenshot 2019-11-20 at 8 20 48 AM" src="https://user-images.githubusercontent.com/31517838/69205113-09170380-0b6f-11ea-9e36-33bc0654bfa2.png">
<img width="1659" alt="Screenshot 2019-11-20 at 7 58 19 AM" src="https://user-images.githubusercontent.com/31517838/69205114-09af9a00-0b6f-11ea-8b16-73521858cb4c.png">
<img width="1667" alt="Screenshot 2019-11-20 at 8 20 01 AM" src="https://user-images.githubusercontent.com/31517838/69205116-09af9a00-0b6f-11ea-8189-9b130067e2b2.png">
<img width="1658" alt="Screenshot 2019-11-20 at 8 19 11 AM" src="https://user-images.githubusercontent.com/31517838/69205117-0a483080-0b6f-11ea-87fb-c1b53c38c018.png">

**Comment by cwebbtw** [comment]
I've experienced similar on all AKS clusters provisioned in Azure.

A workaround has been to set `dnsPolicy: Default`, however with the dns policy set to `ClusterFirst` this is definitely able to be reproduced on 1.13.12.

**Comment by jnoller** [comment]
Please confirm that the OS disk on the worker nodes (core dns runs inside the cluster context) is not IO throttled. For additional information, please see: https://github.com/Azure/AKS/issues/1322#issuecomment-555169820 - I am working on full guidance on the issue.

**Comment by cwebbtw** [comment]
Thanks for your comment @jnoller.

I tested this by provisioning an AKS cluster with DS2_v2 nodes (x3) with only the default pods running on it.

Given I'm restricted to only choosing the machine type, is there a recommended **minimum** machine type that we should be using. I will check the IO asap and feedback.

**Comment by jnoller** [comment]
@cwebbtw  the VM SKU size doesn't actually dictate the OS disk IOPS, that's the *size* of the OS disk. Currently, your cluster has a default 100gb disk (P10 Premium SSD - 100 mbits/sec cap, 500 IOPS quota) - in order to change the disk size you can re-deploy the cluster using the ARM template and overriding the OS disk size setting (working on docs)

**Comment by cwebbtw** [comment]
Thanks @jnoller, I wasn't aware that changing the disk size increased the IOPS. 

I can see from the following capture that the disk size does determine the IOPS.

![image](https://user-images.githubusercontent.com/31657038/69179524-93298280-0b03-11ea-8aa0-680a3c70e49a.png)

I will check to see the current IO usage still.

**Comment by vishnujanardhanan** [comment]
> #1322 (comment)

**Comment by vishnujanardhanan** [comment]
hi @jnoller, stats on OS disk seems to be below the thresholds for P10. Please find the screenshots - stats from 1700-1800 UTC. Time of test 1725-1728.

**Comment by cwebbtw** [comment]
I can confirm that we are also below the thresholds for P10 on the test cluster that I have created.

A quick thought on that matter regarding the workaround that I posted:

> A workaround has been to set dnsPolicy: Default, however with the dns policy set to ClusterFirst this is definitely able to be reproduced on 1.13.12.

If the problem is IO throttling, then setting the dnsPolicy to Default shouldn't fix the problem should it, since DNS queries will still be going directly to the node.

**Comment by adam-yorwerth** [comment]
Can confirm that we're also affected by this issue.

**Comment by jnoller** [comment]
Unless you are collecting the Linux host level IO metrics for the OS disk, and *not* looking at kube metrics, you’re probably throttled.

Unless you override the AKS disk size using arm from the default 100gb on the node the max IOPS quota is 500.

This means an e64v3 sku with say, a peak IOPs on the VM of 75000. If you have a 100gb drive, your max is in reality 500.

The disk stats your views are probably IO and iops reports by kube metrics, this metric is the in memory kube view, not host view

On the host you need to measure io queue depth, bandwidth used - all this would look like from your user perspective is high latency, tunnelfront crashes, and workload instability

Basically, when the throttle kicks in when the cap is hit (as all cloud providers have), the underlying service pushes back. On Linux this causes IO wait to go out of control, but you will not see a spike in CPU as there is no IO going through, it’s all stuck in wait.

A 1ms throttle event will have wide ranging impact as it moves from the kernel/device later up through the full kube stack

**Comment by KrzysztofKowalczyk** [comment]
As we working with containers and don't care about persistence or hosts dying, can AKS be configured to use (and be bound by) the local (temporary) storage not by the NAS drive and it's size/IOPS?

**Comment by GuyPaddock** [comment]
@jnoller Assuming that an existing cluster is encountering this issue, what can be done? You mention customizing the OS disk size via ARM, but does this have to be done when the cluster is being created or can it be done when scaling the cluster?

**Comment by GuyPaddock** [comment]
To answer my own question, it seems you need to complete the following tasks with each node in the cluster:
1. Drain the node.
2. Stop the node from within the Azure portal.
3. Edit the OS disk attached to the node to increase its size.
4. Start the node from within the Azure portal.
5. Un-cordon the node.

In my case, my OS disk was set to 30 GiB (!), which is even worse than the 100 GiB number that was indicated as too low in this thread.

**Comment by jnoller** [comment]
Sorry @GuyPaddock - I’ve been out of office. Yes, that’s the resize process, but new nodes and upgrades will potentially change that back. 

I am working on all info/docs for guidance this week

**Comment by Aanuka** [comment]
Can confirm that we also have this issue with interim 5 seconds delay.
If we loop an curl command (I.e 50 Requests) we can see mostly Low-Response Time with usually 1-2 of 5 seconds delay.

- Latest Kernel (4.15.0-1063-azure)
- 1.14.6
- Azure NPM 1.0.28
- 100GB Default Size
- Low-Load (14% CPU, 45% Mem over the board)
- No I/O Containers (Normal HTTP/Kestrel Workloads based on .NET Container from Microsoft)

We have not yet parsed the I/O Queue Metrics but are rather sure that we are not hitting the limits. In fact, we even scaled the cluster without seeing any affect.

Seeing that it is a 5 seconds delay all of the time, it's almost like the cluster drops the UDP Packets and it has to fallback to TCP. We have no NetworkPolicys in places right now that could do this, nor any NSG that blocks UDP.

**Comment by jnoller** [comment]
@Aanuka  You can not resize your OS disk. The default is 100gb with a peak IOPS of 500.

**Comment by Aanuka** [comment]
@jnoller I'm very aware of that without going through hops of steps.
In our case we scaled the cluster with more nodes and balanced the load, it did not make any effect.
In fact we also tried scaling the cluster and putting the CoreDNS Pods on that machine along with 1 Pod of another workload, it did not make any effect.

If we compare this to another cluster, which has more I/O Intense workloads and on-top of that running Istio (Which our troublesome cluster is not), we can not see the issue on that one.
One big difference between those two is that it's running Kubenet instead of the Azure CNI, which also makes NPM Absolute.

**Comment by jnoller** [comment]
@Aanuka I do think multiple issues are at play here. The kubenet vs CNI is an interesting one, as CNI is more IO heavy on the OS disk than Kubenet is. It is possible it is *also* an issue with NPM.

**Comment by Aanuka** [comment]
@jnoller Agreed. For the time being I can only check what metrics I have, and those give an average of 20 IOPS. As long as the portal aren't showing tremendously wrong, those figures do indicate that we are not near the IOPS Limit of a P10 Disk.

As stated before, we do not have any NetworkPolicys in place on the CNI cluster so uncertain on how that could affect the operations on that one.

**Comment by Aanuka** [comment]
@jnoller We took the time and implemented the same "workaround" seen in #632 
I.e setting "single-request-reopen" with the deployment files.

This directly shaved of 5seconds on ALL Operations if we are looking at the 99th Percentile.

By the looks of it either the #632 issue has been reintroduced or we have something that the same work-around solves.

**Comment by Vandersteen** [comment]
> @jnoller We took the time and implemented the same "workaround" seen in #632
> I.e setting "single-request-reopen" with the deployment files.
> 
> This directly shaved of 5seconds on ALL Operations if we are looking at the 99th Percentile.
> 
> By the looks of it either the #632 issue has been reintroduced or we have something that the same work-around solves.

We also are encountering this issue right now.
Using the 'workaround' helps in our debian based images (asp.net core 2.2), however this does not help in our alpine based images

**Comment by Aanuka** [comment]
> > @jnoller We took the time and implemented the same "workaround" seen in #632
> > I.e setting "single-request-reopen" with the deployment files.
> > This directly shaved of 5seconds on ALL Operations if we are looking at the 99th Percentile.
> > By the looks of it either the #632 issue has been reintroduced or we have something that the same work-around solves.
> 
> We also are encountering this issue right now.
> Using the 'workaround' helps in our debian based images (asp.net core 2.2), however this does not help in our alpine based images

Then it seems to be almost the same issue as #632 with the same caveats around the "workaround" not working on some Alpine images as they don't support the "single-request-reopen" option.

**Comment by SCsnwbrdr** [comment]
I found this thread the other day when a number of our pods were intermittently having a 5s delay as well. 

I created a cron job that would intermittently (once a minute) do an http get request and report the result to app insights. The issue appeared regularly in app insights.

To make sure our environment config was not to blame, I tested a completely fresh aks cluster with the cron job I set up. I noticed that if I set up an aks cluster with my cron job and basic networking the issue was not present. However, when I set up the same test on a fresh aks cluster with advanced networking, the issue occurred regularly with only the cron job deployed to the cluster. Same sized VMs with both clusters.

**Comment by GuyPaddock** [comment]
@jnoller Why did you indicate that the OS disks cannot be resized? Does the procedure I outlined in https://github.com/Azure/AKS/issues/1326#issuecomment-558424615 not apply for some reason?

**Comment by DreamRivulet** [comment]
@cwebbtw  If you use the "Default" policy, then pods will use dns resolve setting on the node, you can check the /etc/resolv.conf on the node, i think it will not use the coredns svc ip as the nameserver.

From my point, this issue may be caused by：1） IO throttling on the node where the coredns pod is running， 2）race issue: https://www.weave.works/blog/racy-conntrack-and-dns-lookup-timeouts
3) others....

**Comment by guitmz** [comment]
This started to happen for me when we upgraded the cluster to v.1.14 (from 1.13). We had the same problem a year ago and it was fixed after the kernel was patched with the mitigations from https://www.weave.works/blog/racy-conntrack-and-dns-lookup-timeouts

is there a way we can run coredns in AKS as daemonset to try to fix this (at least try)? We need azure cni and we cant use `single-request-reopen` for all pods because some are alpine based

**Comment by jnoller** [comment]
Please also see this issue for intermittent nodenotready, DNS latency and other crashed related to system load: https://github.com/Azure/AKS/issues/1373

**Comment by Vandersteen** [comment]
@jnoller Is there a place where we can follow up the resolution of this ? 
From my understanding is that there are 3 bugs in the kernel concerning conntrack race condition.
2 would have been fixed in kernel 4.17.

When can we expect change from AKS ?

**Comment by ridhoq** [comment]
I recently spoke to Azure AKS support about this as we were seeing this behavior in our production cluster.  @Vandersteen mentioned that a fix is available in kernel 4.17, however, the Azure support engineer I worked with reached out to the AKS product team. He said this:

>I got the update from backend team: we didn’t do a test on the kernel 4.17, but as far as we know, the patch on 4.17 did not fix the issue. Since AKS nodes are built with Canonical Ubuntu 16.04 as Linux distribution. Currently 16.04 is using kernel version 4.15.18.  We use a stable Linux distribution to stay with security patches to make sure the security of node ecosystem. You could find more details in this link: https://people.canonical.com/~kernel/info/kernel-version-map.html
However, we do have a plan to update to Ubuntu 18.04, please check this link: https://github.com/Azure/AKS/issues/1487

Long story short, internally, the AKS team thinks that the fix is not related to the version of Ubuntu. If somebody from the AKS team sees this and finds this to be inaccurate, please feel free to chime in

**Comment by mattops** [comment]
Hi.

Thanks for your reply - so do you have any idea when the fix will become
available?

Matt

**Comment by ridhoq** [comment]
Hi @mattops, I did not hear about any ETA or internal milestone when I spoke to Azure Support

**Comment by GuyPaddock** [comment]
In #1373, we're instructed to contact Azure Support to assist with mitigating issues like these. What mitigation steps is support providing in the interim until a workable fix is available? #1373 describes how to observe the problem, and comments from @jnoller tell us to move logs and disable write caching, which is akin to witchcraft until there's some _actual_ documentation from AKS on this issue.

**Comment by Dharm19** [comment]
Is there any update on this from AKS team ? we are seeing this very frequently in our environments

**Comment by timja** [comment]
We deployed nodelocaldns on Friday, and haven't had a single DNS error since doing this, it also greatly improved the performance of our node.js applications as well, (this looks to be the reason: https://medium.com/@amirilovic/how-to-fix-node-dns-issues-5d4ec2e12e95)

n.b. we were unable to get nodelocaldns to work with the azure CNI, but it works with kubenet, we raised this: https://github.com/Azure/AKS/issues/1642

no response from support on it either 😢

**Comment by adinaclaudia** [comment]
@timja How did you manage to deploy nodelocaldns on AKS? I am running into this issue: https://github.com/Azure/AKS/issues/1435 (the daemonset gets deleted automatically after 30-40sec). I have k8s 1.15.10 with kubenet.
My intent was to use nodelocaldns workaround until the feature is implemented in AKS.

**Comment by timja** [comment]
As mentioned in the comments remove the addon manager labels

**Comment by bhicks329** [comment]
We were suffering from random dns timeouts on AKS and turns out our container security software had a bug that identified some of the dns replies as spoofs. Updated to the latest version and all good so far.

**Comment by Berndinox** [comment]
Same issue here: makes AKS literally unusable

**Comment by curtdept** [comment]
ya we are hitting this one

**Comment by davesmits** [comment]
We are facing the same issue and changing the dnsconfig is not working as using a alpine image. If there is no solution we need to migrate to aws...

**Comment by jstewart612** [comment]
Adding report.

**Comment by deyanp** [comment]
When can we finally remove the dnsConfig section from our deployment yamls? Is the Ubuntu 18.04 AKS node image going to remove the need for this workaround?

**Comment by aviramha** [comment]
Facing this too.

**Comment by timja** [comment]
see https://github.com/Azure/AKS/issues/667#issuecomment-706556195

**Comment by palma21** [comment]
Marking these as duplicate/known-issue and adding the in-progress feature for it as well as well as @curtdept current workaround (thanks!)

https://github.com/curtdept/aks_nodelocaldns/blob/main/nodelocaldns.yaml

#1492
